### PR TITLE
add sharded grad and refactor grad hooks

### PR DESCRIPTION
**Pull request by ver217**
Implemented a complete SharedModel FWD/BWD process
This PR uses the newly implemented hook mechanism. 
Specifically, OpHook is used for the management of param.data, and BaseParamHookMgr is used for the reduction and management of param.grad.

### CODE REVIEW

File: colossalai/engine/ophooks/_shard_grad_ophook.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import torch
+from colossalai.registry import OPHOOKS
+
+from . import BaseOpHook
+
+
+@OPHOOKS.register_module
+class ShardGradHook(BaseOpHook):
```
Comment (feifeibear) [inline colossalai/engine/ophooks/_shard_grad_ophook.py:8]: @feifeibear Need a mechanism to guarantee different execution orders. Not related to this PR

File: colossalai/zero/sharded_model/sharded_grad.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                 ) -> None:
+        self.param = param
+        self.process_group = process_group or gpc.get_group(ParallelMode.DATA)
+        self.shard_idx = self.process_group.rank()
+        self.num_shards = self.process_group.size()
+        self.offload_config = offload_config
+
+        self._cpu_offload = offload_config.get('device', None) == 'cpu' if offload_config else False
+
+        sharded_param, _ = get_shard(param.data, self.shard_idx, self.num_shards)
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_grad.py:30]: SharedParam is another Class. You do not need do it here?

File: colossalai/zero/sharded_model/sharded_grad.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        """This function will be called pre-backward
+
+        :raises AssertionError: Raise if grad shape is wrong
+        """
+        if self.param.grad is not None:
+            if self.param.grad.device != self.param.data.device:
+                # TODO: offload?
+                raise RuntimeError(
+                    'grad and param are on different device, grad {self.param.grad.device} vs. param {self.param.data.device}')
+                self.param.grad = None
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_grad.py:56]: Code is a bit redundant. move self.param.grad  = None out of if.
remove usless of branch.

File: colossalai/zero/sharded_model/sharded_grad.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        self._cpu_grad: Optional[torch.Tensor] = None
+
+        if self._cpu_offload:
+            # this buffer will be held and reused every iteration
+            self._cpu_grad = torch.zeros_like(param.data, device='cpu').pin_memory()
+
+    @torch.no_grad()
+    def setup(self) -> None:
+        """This function will be called pre-backward
+
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_grad.py:48]: Need more comments.
Is this function is used to clear the memory space of .grad before BWD?

File: colossalai/zero/sharded_model/sharded_grad.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/zero/sharded_model/sharded_grad.py:30]: Also you shall not assume the parameter.data has been gathered! Therefore, use its size is not appropriate.

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 5]
```diff
@@ -45,19 +60,106 @@ def __init__(self,
...
+        # However, if you set gradient_predivide_factor to None, we will set gradient_predivide_factor to a value >= 1.0 automatically
+        self.gradient_predivide_factor: float = gradient_predivide_factor if gradient_predivide_factor is not None else \
+            get_gradient_predivide_factor(self.world_size)
+        self.gradient_postdivide_factor: float = self.world_size / self.gradient_predivide_factor
+
+        self.comm_stream: torch.cuda.Stream = torch.cuda.Stream()
+        self.reducer = ReduceScatterBucketer(reduce_scatter_bucket_size_mb)
+        self._require_backward_grad_sync: bool = True
+
+        self._register_grad_post_backward_hooks()
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:83]: Duplicated hooks? conflict with the ShardedGradient?

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 6]
```diff
@@ -45,19 +60,106 @@ def __init__(self,
...
+            # Leave the gradient accumulation state as-is if not synchronizing this pass. This ensures p.grad
+            # remains the unsharded gradient accumulated from prior no-sync passes, and _saved_grad_shard
+            # remains the sharded gradient from the last synchronized pass. This also allows interleaved no-sync and
+            # sync passes, if desired.
+            if not self._require_backward_grad_sync:
+                continue
+            p._sharded_grad.write_back()
+
+    @torch.no_grad()
+    def _grad_post_backward_hook(self, param: Parameter, grad: torch.Tensor) -> Optional[torch.Tensor]:
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:116]: do we still need this hook?

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 5 above)
Comment (ver217) [reply colossalai/zero/sharded_model/sharded_model_v2.py:83]: This hook is registered on parameter and it's for gradient reduction. ShardedGradient just manages hooks about setting up gradients before backward.

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 6 above)
Comment (ver217) [reply colossalai/zero/sharded_model/sharded_model_v2.py:116]: In my experiments, I find hooks registered on module will be fired more than once, if the module is used more than once. However, hooks registerd on parameter will be fired only once when gradient is computed, so I think this kind of hook is simpler.

**Review by feifeibear — COMMENTED** [review]
Make sure the shared_model_v2 tests run in real distributed enviroment.

**Review by feifeibear — APPROVED** [review]
LGTM

**Pull request merged**